Test tgmath without printing

tbkka opened the pull request:
The tgmath tests have been comparing the printed form of the numeric results.
This worked only because the old floating-point formatting logic was inaccurate.
As such, they implicitly tolerate slight inaccuracies.

This commit rewrites the test to verify the numeric values directly.
This is both more explicit and more robust.  (In particular, it
protects this test against future changes to the floating-point
formatting logic.)

For example, the previous code verified `log1p` with this code:
```
(d1, f1, g1) = (log1p(dx), log1p(fx), log1p(gx))
expectEqual("log1p 0.0953101798043249 0.0953102 log1p",
  print3("log1p", d1, f1, g1))
```
which implicitly relies on inaccurate printing.

The new code uses a custom helper:
```
expectEqualWithAccuracy(log1p(dx), log1p(fx), log1p(gx), 0.09531017980432487, ulps: 1)
```
which verifies that the `Double`, `Float`, and `CGFloat` values
match the provided number within 1 ULP.

Note: I've somewhat arbitrarily set the `ulps:` parameter to 1 for most tests.
This might be unnecessarily strict and these parameters could be relaxed: The
goal of the tgmath test is simply to ensure that Swift's functions map to the
correct C math library routines.  This test is not trying to validate the
accuracy of those underlying routines.

tbkka committed:
Test tgmath without printing

The tgmath tests have been comparing the printed form of the numeric results.
This worked only because the old floating-point formatting logic was inaccurate.
As such, they implicitly tolerate slight inaccuracies.

This commit rewrites the test to verify the numeric values directly.
This is both more explicit and more robust.  (In particular, it
protects this test against future changes to the floating-point
formatting logic.)

For example, the previous code verified `log1p` with this code:
```
(d1, f1, g1) = (log1p(dx), log1p(fx), log1p(gx))
expectEqual("log1p 0.0953101798043249 0.0953102 log1p",
  print3("log1p", d1, f1, g1))
```
which implicitly relies on inaccurate printing.

The new code uses a custom helper:
```
expectEqualWithAccuracy(log1p(dx), log1p(fx), log1p(gx), 0.09531017980432487, ulps: 1)
```
which verifies that the `Double`, `Float`, and `CGFloat` values
match the provided number within 1 ULP.

Note: I've somewhat arbitrarily set the `ulps:` parameter to 1 for most tests.
This might be unnecessarily strict and these parameters could be relaxed: The
goal of the tgmath test is simply to ensure that Swift's functions map to the
correct C math library routines.  This test is not trying to validate the
accuracy of those underlying routines.

xwu commented:
Does Glibc support a precision of 1 ulp for all tgmath functions? I feel like I’ve definitely had to make tests less stringent for Glibc in my own work.

gottesmm commented:
While you are doing this, you may want to change these tests to use hexfloats instead as well.

stephentyrone commented on test/stdlib/tgmath.swift:
```diff
@@ -18,74 +18,31 @@ import StdlibUnittest
-  }
-#endif
-  return "\(op) \(d) \(f) \(op)"
+func expectEqualWithAccuracy(_ actualDouble: Double, _ expectedDouble: Double, ulps: Int, _ message: String = "", file: String = #file, line: UInt = #line) {
```
I would probably make this a generic on FloatingPoint so we can re-use and update a single definition:
~~~~
internal func equalWithAccuracy<T>(_ observed: T, _ expected: T, ulps: T) -> Bool
  where T: FloatingPoint {
    if (observed == expected) || (observed.isNaN && expected.isNaN) {
      return true
    }
    return abs(observed - expected) <= expected.ulp * ulps
}
~~~~
or similar.

tbkka replied:
I agree your `equalWithAccuracy<T>` would be useful in other tests, but `CGFloat` verification requires extra hoops.  To ensure that `CGFloat` versions match `Double` or `Float` you really need to have all three actuals available.

stephentyrone replied:
I was thinking only of replacing expectEqualWithAccuracy with expectTrue(equalWithAccuracy<T>( ... )) for now.

tbkka replied:
Of course.  Done.

stephentyrone reviewed (commented):
As @xwu noted, 1 ulp is probably too stringent for all platforms, but we can fine-tune the accuracies. This is a "are the right functions getting called" check, rather than a "is the math library good enough" check, so even a much coarser tolerance would be fine.

stephentyrone commented:
@swift-ci please test.

stephentyrone commented on test/stdlib/tgmath.swift:
```diff
@@ -103,200 +60,135 @@ let gz = CGFloat(3.3)
+    expectEqualWithAccuracy(remainder(dx, dy), remainder(fx, fy), remainder(gx, gy), 0.1, ulps: 0)
+    expectEqualWithAccuracy(copysign(dx, dy), copysign(fx, fy), copysign(gx, gy), 0.1, ulps: 0)
+
+    expectEqual(nextafter(dx, dy), 0.10000000000000002)
```
This should be written `expectEqual(nextafter(dx, dy), dx.nextUp)`. This is both more precise and also lets you eliminate the `arch` condition below.

tbkka replied:
Good idea!  I've also added a check that `nextafter()` in the down direction matches `.nextDown`

stephentyrone commented on test/stdlib/tgmath.swift:
```diff
@@ -103,200 +60,135 @@ let gz = CGFloat(3.3)
 
-  (d1, f1, g1) = (nan("12345"), nan("12345"), nan("12345"))
-  expectEqual("nan nan nan nan", print3("nan", d1, f1, g1))
+  expectEqual(remquo(dz, dy).1, 1)
```
The C stdlib `remquo` is a really weird function, and formally only guarantees that the three low-order bits of the quotient are correct. This test should test `remquo(dz, dy).1 & 7`.

tbkka replied:
Thanks for pointing that out.  Done.

tbkka committed:
Verify that C `nextafter` matches Swift `.nextUp`/`.nextDown`

As suggested by Steve Canon.

tbkka committed:
Verify only the bottom three bits of remquo().1

Note: The C/C++ standards only require the bottom three
bits to be correct, although macOS assures 7 bits.

tbkka commented:
@xwu:  I'll take a look at the Linux test results before merging.

As @stephentyrone pointed out, we can easily change these parameters later if Windows or other C libraries require it.

stephentyrone commented on test/stdlib/tgmath.swift:
```diff
@@ -134,13 +134,12 @@ MathTests.test("Binary functions") {
+    expectEqual(nextafter(dx, dy), dx.nextUp)
+    expectEqual(nextafter(fx, fy), fx.nextUp)
+    expectEqual(nextafter(gx, gy), gx.nextUp)
+    expectEqual(nextafter(dx, dx - 1.0), dx.nextDown)
```
You shouldn't actually need `Float(1.0)` here; this can just be `fx - 1`.

stephentyrone commented on test/stdlib/tgmath.swift:
```diff
@@ -174,9 +173,9 @@ MathTests.test("Other functions") {
-  expectEqual(remquo(dz, dy).1, 1)
-  expectEqual(remquo(fz, fy).1, 1)
-  expectEqual(remquo(gz, gy).1, 1)
+  expectEqual(remquo(dz, dy).1 & 7, 1)
```
Please add a comment explaining where the `& 7` comes from. It can be just the sentence in my review note.

tbkka committed:
Comment and simplify a bit

stephentyrone commented:
@swift-ci Please test

stephentyrone commented:
Deleted some noise comments from builds before last round of changes.

tbkka commented:
For the record, ±1 ulp appears to be sufficient tolerance for glibc:
```
14:03:36 PASS: Swift(linux-x86_64) :: stdlib/tgmath.swift (7113 of 11058)
```

tbkka commented:
@gottesmm -- I personally find hex floats appropriate when you want to verify or specify a particular bit pattern.  For cases like this, I find accurate decimal forms easier to read and understand.

stephentyrone commented:
@xwu Are you OK with merging this as is, but adding some comments that clarify that these tolerances can be loosened basically as needed on a follow-on commit? The existing tests expect `.description` to be identical, which is even worse, so this is already an improvement.

xwu commented:
Seems reasonable to me!
